### DISCUSSION TITLE
Parcel help

### DISCUSSION THREAD

**Discussion by IMGROOT2** [For-0/vocabustudy#177] [Q&A] [answered]
I saw that you're using parcel to run the code via localhost, but I'm having trouble with parcel. When I open up localhost:1234, I end up with only the index.pug loaded, but when I try to click on the privacy policy, I'm getting redirected back to index.html. I know it's a problem with the file links, but I wanted to know two things.

1. What do you use to code? Does it directly support pug, or did you have to set up a file watcher to translate it into HTML?
2. Have you ever ran into my issue before?

ty for the help, I'll start coding when I get my end working

**Comment by grimsteel** [comment, marked as answer]
I think you're running the parcel dev server.
For the best emulation, I also use the Firebase Emulator, which includes Auth, Firestore, and Hosting.
Instead of running just `npm run start`, run `npm run watch`
This will ONLY compile the code with hot reloading.
To run the server, also run `npm run emulator`

Before you start the emulator, you need to install firebase tools.
They have some issues in the latest version regarding `/` serving `/index.html`, so install 11.13.0 instead
`npm install -g firebase-tools@11.13.0`
Thanks!

**Reply by grimsteel** [reply]
The Firebase Hosting emulator is what rewrites "/privacy-policy.html" to "/privacy/"

**Reply by grimsteel** [reply]
That `-g` flag in the npm install firebase tools is important because it installs it globally instead of in the local package

**Reply by grimsteel** [reply]
If everything goes, well you should be able to access the mock console at localhost:4000 and the emulated site at localhost:5000

**Reply by grimsteel** [reply]
https://github.com/orgs/For-0/teams/vocabustudy-developers/discussions/3

**Reply by grimsteel** [reply]
Also, what do you mean by this:

> What do you use to code? Does it directly support pug

Are you asking about an IDE, build tool, etc?

**Reply by IMGROOT2** [reply]
IDE. I have one more issue.

I followed the instructions, and the emulator is working well. But, "localhost:5000" redirects to the 404 page, "localhost:5000/#" also redirects to the 404 page, but only when I manually type in "localhost:5000/index.html" does it actually go to the home page. "localhost:5000/privacy/" works fine too. Any idea what's wrong?

**Reply by grimsteel** [reply]
> I saw that you're using parcel to run the code via localhost

The main feature of Parcel is a build tool - something that will transpile the pug, JS w/ module imports, and scss into HTML, Vanilla JS, and CSS

**Reply by grimsteel** [reply]
> I followed the instructions, and the emulator is working well. But, "localhost:5000" redirects to the 404 page, "localhost:5000/#" also redirects to the 404 page, but only when I manually type in "localhost:5000/index.html" does it actually go to the home page. "localhost:5000/privacy/" works fine too. Any idea what's wrong?

That's an issue in the latest version of `firebase-tools`
Make sure you install verion 11.13.0

**Reply by grimsteel** [reply]
firebase/firebase-tools#5133

**Reply by grimsteel** [reply]
> IDE

VS Code, but that shouldn't have anything to do with the building/serving

Edit: I don't even think VS Code is an IDE

**Reply by grimsteel** [reply]
Also, assuming you are using some kind of BASH shell, run this to make sure that the emulator-data does not get committed to Source Control (I want it to be in the repo, but I only want one copy of it)
`git update-index --assume-unchanged emulator-data/**/*.*`
Before you run that, make sure that `globstar` is enabled (`shopt -s globstar` to enable it - you can disable it with `shopt -u globstar` afterwards)

**Reply by IMGROOT2** [reply]
Thanks! I think I installed the latest version by accident while trying to clear a vulnerability lol

**Reply by grimsteel** [reply]
Yes same...
Did auditing work for you globally?
I just did the thing where you init a "package" in the global folder, then audit from there

**Reply by IMGROOT2** [reply]
> Also, assuming you are using some kind of BASH shell, run this to make sure that the emulator-data does not get committed to Source Control (I want it to be in the repo, but I only want one copy of it) `git update-index --assume-unchanged emulator-data/**/*.*` Before you run that, make sure that `globstar` is enabled (`shopt -s globstar` to enable it - you can disable it with `shopt -u globstar` afterwards)

I'll just only commit changes to the actual files (by that I mean, only the .pug, .scss, etc.)

**Reply by IMGROOT2** [reply]
> Yes same... Did auditing work for you globally? I just did the thing where you init a "package" in the global folder, then audit from there

I don't think so

**Reply by IMGROOT2** [reply]
![image](https://user-images.githubusercontent.com/116324098/197417116-2443ca33-86fa-4371-8cd0-d201784fd797.png)

I won't be committing any of these for example

**Reply by grimsteel** [reply]
Yes, thanks!
If you want, you can add `.idea/` to the `.gitignore` - I added `.vscode/` to it, so that's fair
But as for the emulator data, I want it to be included in the repo, I just don't want to keep commiting those files every time. (However, I just made some critical changes to it, so I'll probably commit it this time)
